contracts/stake-cw20-external-rewards/src/contract.rs

```diff
@@ -0,0 +1,1324 @@
+    let mut amount = Uint128::zero();
+    while current_block <= min(up_to_block, config.end_block) {
+        let stake_info = get_stake_info_at_height(deps.querier, config, address, current_block)?;
+        if stake_info.balance > Uint128::zero() {
+            amount += ((Uint128::from(delta) * config.payment_per_block) * stake_info.balance)
+                .checked_div(stake_info.total)?
+        }
+        current_block += delta;
+    }
+    Ok((amount, current_block))
+}
+
+fn get_payment_msg(
+    _env: &Env,
+    config: &Config,
+    info: &MessageInfo,
+    amount: Uint128,
+) -> StdResult<CosmosMsg> {
+    Ok(match config.clone().denom {
+        Denom::Native(denom) => BankMsg::Send {
+            to_address: info.sender.to_string(),
+            amount: vec![Coin { amount, denom }],
+        }
+        .into(),
+        Denom::Cw20(address) => {
+            let transfer = Cw20ExecuteMsg::Transfer {
+                recipient: info.sender.to_string(),
+                amount,
+            };
+            WasmMsg::Execute {
+                contract_addr: address.to_string(),
+                msg: to_binary(&transfer)?,
+                funds: vec![],
+            }
+            .into()
+        }
+    })
+}
+
+struct StakeInfo {
+    pub balance: Uint128,
+    pub total: Uint128,
+}
+fn get_stake_info_at_height(
+    deps: QuerierWrapper,
+    config: &Config,
+    address: &Addr,
+    height: u64,
+) -> StdResult<StakeInfo> {
+    let balance_query = stake_cw20::msg::QueryMsg::StakedBalanceAtHeight {
+        address: address.to_string(),
+        height: Some(height),
+    };
+    let balance_response: stake_cw20::msg::StakedBalanceAtHeightResponse =
+        deps.query_wasm_smart(config.staking_contract.to_string(), &balance_query)?;
+    let total_query = stake_cw20::msg::QueryMsg::TotalStakedAtHeight {
+        height: Some(height),
+    };
+    let total_response: stake_cw20::msg::TotalStakedAtHeightResponse =
+        deps.query_wasm_smart(config.staking_contract.to_string(), &total_query)?;
+    Ok(StakeInfo {
+        balance: balance_response.balance,
+        total: total_response.total,
+    })
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::Info {} => to_binary(&query_info(deps)?),
+        QueryMsg::ClaimableRewards { address } => {
+            to_binary(&query_claimable_rewards(deps, env, address)?)
+        }
+    }
+}
+
+pub fn query_info(deps: Deps) -> StdResult<InfoResponse> {
+    let config = CONFIG.load(deps.storage)?;
+    Ok(InfoResponse {
+        start_block: config.start_block,
+        end_block: config.end_block,
+        payment_per_block: config.payment_per_block,
+        total_amount: config.total_amount,
+        denom: config.denom,
+        staking_contract: config.staking_contract.to_string(),
+        payment_block_delta: config.payment_block_delta,
```

Comment:
Might be nice to include if the contract has been funded in here.

Reply:
good catch